api-parent/core-api/src/main/java/com/sap/cloud/environment/servicebinding/api/ServiceBinding.java

```diff
@@ -99,4 +100,161 @@ default Optional<ServiceIdentifier> getServiceIdentifier()
      */
     @Nonnull
     Map<String, Object> getCredentials();
+
+    /**
+     * Returns a <b>NEW</b> {@link ServiceBinding} instance that contains the same values as this {@link ServiceBinding}
+     * but with the given {@code properties}.
+     * <p>
+     * <b>Note</b>: The given {@code properties} will affect {@link #getKeys()}, {@link #containsKey(String)}, and
+     * {@link #get(String)} <b>only</b>. All other methods will return the exact same values as this
+     * {@link ServiceBinding}.
+     * </p>
+     * <p>
+     * Example:
+     *
+     * <pre>
+     * ServiceBinding originalBinding;
+     * originalBinding.getKeys(); // ["old-key1", "old-key2"]
+     *
+     * ServiceBinding bindingWithArbitraryKey =
+     *     originalBinding.withProperties(Collections.singletonMap("new-key", "some value"));
+     * bindingWithArbitraryKey.getKeys(); // ["new-key"]
+     *
+     * originalBinding.getCredentials(); // {"key-1": "value-1", "key-2": "value-2"}
+     * ServiceBinding bindingWithCredentialsProperty =
+     *     originalBinding.withProperties(Collections.singletonMap("credentials", "some value"));
+     * bindingWithCredentialsProperty.getCredentials(); // {"key-1": "value-1", "key-2": "value-2"} <-- this has NOT been changed
+     * </pre>
+     * </p>
+     *
+     * @param properties
+     *            The properties to set. Supplying {@code null} is equivalent to supplying an empty {@link Map}.
+     * @return A <b>NEW</b> {@link ServiceBinding} instance that contains the same values as this {@link ServiceBinding}
+     *         but with the given {@code properties}.
+     * @since 0.10.0
+     */
+    @Nonnull
+    default ServiceBinding withProperties( @Nullable final Map<String, Object> properties )
+    {
+        return DelegatingServiceBinding.builder(this).withProperties(properties).build();
+    }
```

Comment:
I feel it's redundant to have both, a wither pattern and a builder pattern 🤔

For an expert library like this, I would expect API minimalism and no usage ambiguity.

Reply:
Please note that the `DelegatingServiceBinding` (including it's builder) is **not** public API.
The customers cannot (as of now) use the `DelegatingServiceBinding.builder(...)` themselves and always have to rely on the `ServiceBinding.with<Foo>(...)` API.

This decision can (and should) be challenge IMO.

As an alternative, for example, we could drop the additional API on the `ServiceBinding` interface (to keep the API as slim as possible) and just make the `DelegatingServiceBinding` class public.
This would come with the downside of more boilerplate code for "manipulating" individual properties:

```java
ServiceBinding originalBinding;
ServiceBinding changedBinding = originalBinding.withName("new name");

vs.

ServiceBinding originalBinding;
ServiceBinding changedBinding = DelegatingServiceBinding.builder(originalBinding).withName("new name").build();
```

Reply:
> the `DelegatingServiceBinding` (including it's builder) is **not** public API.

Could we remove the API then in favor of _wither_ pattern?

Reply:
This library does not contain `Lombok` (to keep the dependency footprint as small as possible).

---

> Could we remove the API then in favor of wither pattern?

Sure. We could remove the `DelegatingServiceBinding.Builder` class, but then we would need to call the constructor, which is easy to mess up due to all the `null`s that need to be passed.

Reply:
Instead of builder pattern, would / have you considered a mutable class with constant delegates and nullable attributes that have a package private setter?

You could still instantiate a new DelegatingServiceBinding for every wither.

Reply:
(We talked about the motivation and actual customer use-case.)

The original issue is that the default implementation of `getServiceIdentifier()` serves us wrong, in scenarios we didn't expect. Our assumption `/service name/ -> /service identifier/` is proven wrong!

And enabling additional API to provide better heuristic(s) to infer `/service identifier/` will be too big of a scope at the moment. Therefore we'll allow for delegating/manipulating copies of `ServiceBinding` instead. This will enable the customer use-case again and buys us time to rethink the logic.